Only keep last 10 session logs.

bd82 opened the pull request:
Signed-off-by: I060847 <shachar.soel@sap.com>

#### What it does
Only keeps at most `N` most recent session log folders.
- N is configurable using the `--plugin-max-session-logs-folders=N` CLI option.
- N is 10 by default as is the behavior in VSCode.

Fixes #6866

#### How to test

- Start examples/browser
- Open the theia logs folder
  - on windows: `C:\Users\[USER-NAME]\AppData\Roaming\.theia\logs`	
- Open `http://localhost:3000` and refresh a dozen times.
- Note the theia logs folder only contains 10 timestamp sub folders
  and that those are the most recent timestamps.

- Stop theia.
- Start theia using `--plugin-max-session-logs-folders=2` CLI option.
- Open `http://localhost:3000`.
- Open the theia logs folder
- Inspect there are only 2 timestamp sub folders there.

I do not know if and how this could be tested as part of an integration test.
Nor do I think I have any more time to implement such a test 😢 ..

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

lmcbout commented on packages/plugin-ext/package.json:
```diff
@@ -78,6 +79,7 @@
     "@theia/ext-scripts": "^0.14.0",
     "@types/decompress": "^4.2.2",
     "@types/escape-html": "^0.0.20",
+    "@types/fs-extra": "^4.0.2",
```
Why do you add "@types/fs-extra": "^4.0.2", in devDependencies when you already included it in dependencies ?
Not needed here.

marechal-p replied:
I don't see the duplication. Typings are added as devdeps, with the actual thing as a dep.

bd82 replied:
Hello.

The fs-extra package does not include type signatures. Therefore the respective @types package is consumed as a dev dep.

Fs extra is already at version 8. However i did not want to introduce a new dep as part if this change so i used a version which is already consumed else where in theia.

bd82 replied:
Hello.

The fs-extra package does not include type signatures. Therefore the respective @types package is consumed as a dev dep.

Fs extra is already at version 8. However i did not want to introduce a new dep as part if this change so i used a version which is already consumed else where in theia.

marechal-p commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -95,7 +99,14 @@ export class PluginPathsServiceImpl implements PluginPathsService {
+        const timeStamp = new Date().toISOString().replace(/[-:]|(\..*)/g, '');
+        // Ensures our timestamp generation logic is "valid".
+        // Changes to the timestamp structure may break old logs deletion logic.
+        debugger;
```
Left over?

bd82 replied:
Oops i was sure i deleted this :(

bd82 replied:
Maybe i did not force push latest fix. Will check when next to laptop again

marechal-p commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -17,13 +17,17 @@
 import { THEIA_EXT, VSCODE_EXT, getTemporaryWorkspaceFileUri } from '@theia/workspace/lib/common';
 
+const MAX_LOG_SESSIONS_TO_KEEP = 10;
+const SESSION_TIMESTAMP_PATTERN = /\d\d\d\d\d\d\d\dT\d\d\d\d\d\d/;
```
```ts
const SESSION_TIMESTAMP_PATTERN = /\d{8}T\d{6}/;
```
Easier to read?

bd82 replied:
Yes. I will fix this later

bd82 replied:
changed to: `/^\d{8}T\d{6}$/` so it would only match **exact* length matches.
`

bd82 commented on examples/browser/webpack.config.js:
```diff
@@ -1,18 +1,165 @@
-/**
- * This file can be edited to customize webpack configuration.
```
how come did i change this file !?

bd82 commented:
O.k. ready for re-review 😄

vince-fugnitto commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -17,13 +17,17 @@
 import { injectable, inject } from 'inversify';
 import { FileSystem, FileStat } from '@theia/filesystem/lib/common';
 import * as path from 'path';
+import {  readdir, remove} from 'fs-extra';
```
I'm not sure why it wasn't picked up by our formatting rules, but there seems to be an extra leading `whitespace`.
If I format the file the content gets updated (which means anyone in the future will have this line as a change).

bd82 replied:
btw I would suggest evaluating using prettier for formatting and a linter only for semantic errors. You can setup a auto formatter on a pre-commit hook using "husky"
and never worry about formatting issues in review ever again.

- https://prettier.io/

vince-fugnitto replied:
Yes, I've seen husky before, I think the fact we haven't set up something like it yet is due to priority but it certainly looks like a nice tool to have.

bd82 replied:
Here is an example package.json with:

- prettier
- husky pre-commit hook for prettier
- Conventional Angular Commit messages
- Automatic changeLog generation (per package).

https://github.com/SAP/xml-tools/blob/master/package.json

BTW: if you ever integrate prettier you will need to use

- https://github.com/prettier/tslint-config-prettier

to disable all syntax formatting rules in TS-Lint to avoid conflicts.

Perhaps this would be of interest to you...

vince-fugnitto commented on CHANGELOG.md:
```diff
@@ -78,6 +78,7 @@ Breaking changes:
   If it does not have a reference to `gen-webpack.config.js` then it will be regenerated.
 - [debug] removed `@theia/json` dependency. Applications should explicitly depend on `@theia/json` instead [#6647](https://github.com/eclipse-theia/theia/pull/6647)
 - [preferences] removed `@theia/json` dependency. Applications should explicitly depend on `@theia/json` instead [#6647](https://github.com/eclipse-theia/theia/pull/6647)
+- [plugin-ext] only keep at most 10 last session logs sub-folders.
```
[minor] do you mind updating this entry?
1. the entry should not be placed under `Breaking Changes`
2. the entry should reference the corresponding pull-request like other entries
3. it should be updated to something more descriptive, for example:
```md
- [plugin-ext] added functionality to remove session logs from sub-folders older than 10 sessions [#6956](https://github.com/eclipse-theia/theia/pull/6956)
```

bd82 replied:
Sure

vince-fugnitto commented on packages/plugin-ext/package.json:
```diff
@@ -29,6 +29,7 @@
     "connect": "^3.7.0",
     "decompress": "^4.2.0",
     "escape-html": "^1.0.3",
+    "fs-extra": "^4.0.2",
```
There is no need to add `fs-extra` or `@types-fs-extra` since they are pulled from `@theia/core` (the root extension for all extensions). There is an additional `tsconfig` rule to make dependencies exceptions for the `no-implicit-dependency` rule:

https://github.com/eclipse-theia/theia/blob/9fb6351621583f4a3e51954debbea9ed1e080168/packages/core/package.json#L14

https://github.com/eclipse-theia/theia/blob/9fb6351621583f4a3e51954debbea9ed1e080168/packages/core/package.json#L29

https://github.com/eclipse-theia/theia/blob/9fb6351621583f4a3e51954debbea9ed1e080168/configs/errors.tslint.json#L59

The corresponding pull-request: https://github.com/eclipse-theia/theia/commit/f9ff23786ed48f611aa335527c78db70427edc54

bd82 replied:
done

vince-fugnitto commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -95,7 +99,13 @@ export class PluginPathsServiceImpl implements PluginPathsService {
      * Generate time folder name in format: YYYYMMDDTHHMMSS, for example: 20181205T093828
      */
     private gererateTimeFolderName(): string {
```
[minor] since you are already updating the following method, do you mind renaming the method to fix the typo:
```ts
generateTimeFolderName()
```

bd82 replied:
done

vince-fugnitto commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -95,7 +99,13 @@ export class PluginPathsServiceImpl implements PluginPathsService {
+        // Ensures our timestamp generation logic is "valid".
+        // Changes to the timestamp structure may break old logs deletion logic.
+        if (!SESSION_TIMESTAMP_PATTERN.test(timeStamp)) {
+            throw new Error(`Failed Generating Session TimeStamp: <${timeStamp}>`);
```
The old implementation would simply return a `string`, but now clients can potentially be thrown an error.
Should we update clients to perform error handling?

akosyakov replied:
or don't throw, just log an error?

bd82 replied:
The intent here was not to expend the methods API to include `@throws`
Rather it was to ensure that the **created** folderNames are **aligned** with the **deleted** folder names. as a mis-alignments will break the cleanup feature...

This is sort of like a compiler check at runtime... I want to "break everything"
If somebody breaks the alignment, however I am not sure a "throws" will suffice.
So nobody can **merge** changes to master that break this alignment.

So what should I do here?
- Some hacky `process.exit(666)` to ensure breakage?
- Is there some other way to halt Theia with a meaningful error message?
- Logging and hoping somebody will look at the logs? (won't prevent merging code that causes mis-alignment).

Thanks.
Shahar.

benoitf replied:
logging should be fine IMHO

bd82 replied:
I've changed it to log an error

akosyakov commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -17,13 +17,17 @@
 import { injectable, inject } from 'inversify';
 import { FileSystem, FileStat } from '@theia/filesystem/lib/common';
 import * as path from 'path';
+import { readdir, remove} from 'fs-extra';
```
missing whitespace, i though we have format on save enabled. which ide do you use?

bd82 replied:
I use Jetbrains products, we are really wasting everybody's time by not having a CLI level formatting solution...

I can see if I have time next week to make a PR with such a configuration, however it will re-format the entire project forcing many re-bases which means somebody from the core team will have to take it farther... (I can make a "pure" PR with the configuration only...)

Are you interested in that?

bd82 replied:
I fixed the WS

akosyakov commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -42,7 +46,7 @@ export class PluginPathsServiceImpl implements PluginPathsService {
         const pluginDirPath = path.join(parentLogsDir, this.gererateTimeFolderName(), 'host');
         await this.fileSystem.createFolder(pluginDirPath);
-
+        await this.cleanupOldLogs(parentLogsDir);
```
How often is this method called? Would it slow it down?

bd82 replied:
It gets called twice for some reason, the remove of each folder is intentionally not "awaited" to. however each call does scan the sub-folder names of the root log folder, so that could add some time.

Note that it is strange that it is called twice, because if two calls are done in **different seconds** then two log session folders could be created with timestamp different by 1 second.

benoitf replied:
we should never wait on cleanup

bd82 replied:
`this.cleanupOldLogs(parentLogsDir)` is now called without `await`
To avoid waiting on cleanup.

benoitf commented on packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts:
```diff
@@ -17,13 +17,17 @@
 import { PluginPathsService } from '../../common/plugin-paths-protocol';
 import { THEIA_EXT, VSCODE_EXT, getTemporaryWorkspaceFileUri } from '@theia/workspace/lib/common';
 
+const MAX_LOG_SESSIONS_TO_KEEP = 10;
```
Hello, this hardcoded should be configurable

bd82 replied:
If you can point me to a quick win example how to expose configurable options in Theia. I can have a look (this is my first contribution to theia and I do not know my way around well yet...).

If the configuration is not trivial, I would prefer to keep it out of this PR
and start with alignment to VSCode implementation of keeping 10 **hard-coded**
last sessions.
- https://github.com/Microsoft/vscode/blob/613447d6b3f458ef7fee227e3876303bf5184580/src/vs/code/electron-browser/sharedProcess/contrib/logsDataCleaner.ts#L32

Anyhow in a **Web IDE** scenario we may want a more advanced configuration
assuming disc space on the containers is a somewhat costly resource:
e.g: Max N Folders **AND** max Y logs size in megabytes.
So the configuration option may need to be "attacked" **separately** regardless.

benoitf replied:
hello @bd82, here is an example for 'plugins' which can be configured through cli with 'plugins' arg

bd82 replied:
I think the link is missing,
But is something like this what you are referring to?

- https://github.com/eclipse-theia/theia/blob/6347c5ccbf016b7cfa70976f69616eb1bb49c5b0/packages/plugin-ext/src/main/node/plugin-deployer-impl.ts#L45-L46

benoitf replied:
https://github.com/eclipse-theia/theia/blob/615e817b68e114531282a8a0af34c5af08c573c7/packages/plugin-ext/src/main/node/plugin-cli-contribution.ts

Sorry it was this link yes

bd82 replied:
Done

bd82 commented:
Read for re-review.

akosyakov commented on packages/plugin-ext/src/main/node/plugin-cli-contribution.ts:
```diff
@@ -23,8 +23,13 @@ import { LocalDirectoryPluginDeployerResolver } from './resolvers/plugin-local-d
 
     static PLUGINS = 'plugins';
+    static MAX_SESSION_LOGS_FOLDERS = 'max-session-logs-folders';
+    // This is the default value used in VSCode, see:
```
/** */ - for member js-doc

akosyakov commented on packages/plugin-ext/src/main/node/plugin-cli-contribution.ts:
```diff
@@ -23,8 +23,13 @@ import { LocalDirectoryPluginDeployerResolver } from './resolvers/plugin-local-d
 export class PluginCliContribution implements CliContribution {
 
     static PLUGINS = 'plugins';
+    static MAX_SESSION_LOGS_FOLDERS = 'max-session-logs-folders';
```
please prefix with plugin, Theia has primary logs and this does affect them

bd82 replied:
done

akosyakov commented:
@vince-fugnitto @benoitf would you be able to finish the review please

bd82 commented:
I fixed @akosyakov minor comments and went over the code
and fixed another minor whitespace formatting issue.

Can we please move this PR forward ? 😄 

Thanks.
Shahar.

akosyakov commented:
It’s fine with me, but someone has to test

vince-fugnitto commented:
> Can we please move this PR forward ? 😄

I plan on testing, can you please fix the CI there are failures.

bd82 commented:
travis build is failing due to two tslint errors of ignoring `any`.
```
/home/travis/build/eclipse-theia/theia/packages/plugin-ext/src/main/node/paths/plugin-paths-service.ts
  153:58  error  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  155:56  error  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
```

It does not look like the ts-lint ignore comment is applied correctly.
```typescript
        // tslint:disable-next-line:no-any
        const subDirEntries = dirEntries.filter((dirent: any) => dirent.isDirectory() );
        // tslint:disable-next-line:no-any
        const subDirNames = subDirEntries.map((dirent: any) => dirent.name);
```

any ideas?

bd82 commented:
I am trying:
```typescript
        // tslint:disable:no-any
        const subDirEntries = dirEntries.filter((dirent: any) => dirent.isDirectory() );
        const subDirNames = subDirEntries.map((dirent: any) => dirent.name);
        // tslint:enable:no-any
```

vince-fugnitto commented:
> any ideas?

The following worked for me:

```ts
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const subDirEntries = dirEntries.filter((dirent: any) => dirent.isDirectory());

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const subDirNames = subDirEntries.map((dirent: any) => dirent.name);
```

bd82 commented:
Thanks @vince-fugnitto 

I've applied the first solution you proposed, lets hope it works now 😄

vince-fugnitto commented:
> Thanks @vince-fugnitto
> 
> I've applied the first solution you proposed, lets hope it works now 😄

It passed :) testing now...

bd82 commented:
> Does the issue successfully close #6866?

Possibly, depending if you want even more configurable options or not, e.g: max session logs in megabytes...

> My only minor complaint would be the commit title + message in order to better describe the changes part of the pull-request. We should describe what the change does and also the fact that we mention 10 last session logs is incorrect now that we have a cli argument which can be passed.

Yes that is true, this is due to increasing the scope of the PR, I only intended to keep the last 10 (**hard-coded**) sessions as VSCode does.
However, `feature creep` increased the scope of this PR...

Anyhow I think this can easily be fixed if you **squash** the commit and **modify** the message... instead of waiting another hour for travis build...

Let me know if this does not work and I will update it on my branch...

vince-fugnitto commented:
> > Does the issue successfully close #6866?
> 
> Possibly, depending if you want even more configurable options or not, e.g: max session logs in megabytes...

I think the base functionality is now present in this pull-request so I'm happy to close it :)
If we think of additional options, or users ask then we can open corresponding issues to address the improvements on a case-by-case basis.

> Yes that is true, this is due to increasing the scope of the PR, I only intended to keep the last 10 (**hard-coded**) sessions as VSCode does.
> However, `feature creep` increased the scope of this PR...

That's not a problem :)

> Anyhow I think this can easily be fixed if you **squash** the commit and **modify** the message... instead of waiting another hour for travis build...

In general that is not the workflow present in the project, we ask that authors usually write the commits and pull-requests completely, there is no post-processing done after the fact. We also use a `rebase and merge` approach when merging pull-requests. 

Please feel free to update the commit message, I think waiting for the Travis build will be fine, the pull-request will be merged today which will definitely be in the release tomorrow so there are no worries :)

bd82 commented:
@vince-fugnitto 

Commit message and PR description have been updated please inspect.

### Regarding feature creep:
It is not a problem if there is "unlimited" time. However if somebody wants to contribute a small feature and has only very little time to attempt so, particularly if that someone is a **first time contributor** and need to pass other hurdles first (dev env / Inversify / Legal) I believe **the smaller the scope the better**, particularly
as the smaller scope in this case is identical to how VSCode does things...

Nothing in this PR is really complicated, it just took more time then I thought
it would... and some of that extra time could have been avoided...

Cheers.
Shahar.

vince-fugnitto commented:
> @vince-fugnitto
> 
> Commit message and PR description have been updated please inspect.
> 
> ### Regarding feature creep:
> It is not a problem if there is "unlimited" time. However if somebody wants to contribute a small feature and has only very little time to attempt so, particularly if that someone is a **first time contributor** and need to pass other hurdles first (dev env / Inversify / Legal) I believe **the smaller the scope the better**, particularly
> as the smaller scope in this case is identical to how VSCode does things...
> 
> Nothing in this PR is really complicated, it just took more time then I thought
> it would... and some of that extra time could have been avoided...
> 
> Cheers.
> Shahar.

I understand your concerns, please note that in any project (especially as a first-time contributor), there are always steps required such a setting up a dev environment, understanding the codebase, testing, and validation. Things take time and should be properly completed before being accepted.

I also understand that the configurable part of the pull-request could have probably have waited as an enhancement after this pull-request was merged, but I also believe I mentioned the commit message during my initial review https://github.com/eclipse-theia/theia/pull/6956#pullrequestreview-347677214.

vince-fugnitto commented:
Thank you for your contribution @bd82 I appreciate it! 👍

bd82 commented:
> Thank you for your contribution @bd82 I appreciate it! 👍

Happy to help 😄
